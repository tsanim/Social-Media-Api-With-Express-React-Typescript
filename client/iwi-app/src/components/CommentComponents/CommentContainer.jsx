import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHeart, faTrash } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types';

function CommentContainer(props) {
    return (
        <div className="commentContainer">
            {/* If edit form is shown , both comment's text and buttons are not shown */}
            <p>{props.text} </p>
            <div className="commentBtns">
                {
                    props.isLiked
                        ? <button onClick={props.handleDisLikeComment} className="liked"><FontAwesomeIcon icon={faHeart} /></button>
                        : <button onClick={props.handleLikeComment}><FontAwesomeIcon icon={faHeart} /></button>
                }

                {
                    (props.currentUser._id ? props.currentUser._id : props.currentUser.id) === props.commentCreator._id
                        ? <button onClick={props.handleShowModal} ><FontAwesomeIcon icon={faTrash} /></button>
                        : null
                }
            </div>
        </div>
    )
}

CommentContainer.propTypes = {
    text: PropTypes.string,
    isLiked: PropTypes.bool,
    currentUser: PropTypes.object,
    commentCreator: PropTypes.object,
    handleDisLikeComment: PropTypes.func,
    handleLikeComment: PropTypes.func,
    handleShowModal: PropTypes.func
}

export default CommentContainer;