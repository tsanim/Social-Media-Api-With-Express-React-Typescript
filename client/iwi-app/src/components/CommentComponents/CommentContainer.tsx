import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHeart, faTrash } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types';
import CommentContainerProps from '../../interfaces/Components/CommentComponents/CommentContainerProps.interface';

function CommentContainer(props: CommentContainerProps) {
    return (
        <div className="commentContainer">
            {/* If edit form is shown , both comment's text and buttons are not shown */}
            <p>{props.text} </p>
            <div className="commentBtns">
                {
                    props.isLiked
                        ? <button onClick={props.handleDislikeComment as () => void} className="liked"><FontAwesomeIcon icon={faHeart} /></button>
                        : <button onClick={props.handleLikeComment as () => void}><FontAwesomeIcon icon={faHeart} /></button>
                }

                {
                    (props.currentUser._id ? props.currentUser._id : props.currentUser.id) === props.commentCreator._id
                        ? <button onClick={props.handleShowModal} ><FontAwesomeIcon icon={faTrash} /></button>
                        : null
                }
            </div>
        </div>
    )
}

CommentContainer.propTypes = {
    text: PropTypes.string,
    isLiked: PropTypes.bool,
    currentUser: PropTypes.object,
    commentCreator: PropTypes.object,
    handleDisLikeComment: PropTypes.func,
    handleLikeComment: PropTypes.func,
    handleShowModal: PropTypes.func
}

export default CommentContainer;