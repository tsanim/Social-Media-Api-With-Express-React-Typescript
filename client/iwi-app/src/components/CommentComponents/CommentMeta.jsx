import React from 'react';
import calcTime from '../../utils/calcTime';
import PropTypes from 'prop-types';

function CommentMeta(props) {
    return (
        <div className="meta">
            <span className="date">{calcTime(props.date)}</span>
            <span onClick={props.handleShowLikesModal} >{props.likesCount} likes</span>
        </div>
    )
}

CommentMeta.propTypes = {
    date: PropTypes.string,
    likesCount: PropTypes.number,
    handleShowLikesModal: PropTypes.func,
    testProp: PropTypes.string
}

export default CommentMeta;