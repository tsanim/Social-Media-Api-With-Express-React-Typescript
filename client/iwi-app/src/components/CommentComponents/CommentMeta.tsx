import React from 'react';
import calcTime from '../../utils/calcTime';
import PropTypes from 'prop-types';
import CommentMetaProps from '../../interfaces/Components/CommentComponents/CommentMetaProps.interface';

function CommentMeta(props: CommentMetaProps) {
    return (
        <div className="meta">
            <span className="date">{calcTime(props.date)}</span>
            <span onClick={props.handleShowLikesModal} >{props.likesCount} likes</span>
        </div>
    )
}

CommentMeta.propTypes = {
    date: PropTypes.string,
    likesCount: PropTypes.number,
    handleShowLikesModal: PropTypes.func,
}

export default CommentMeta;