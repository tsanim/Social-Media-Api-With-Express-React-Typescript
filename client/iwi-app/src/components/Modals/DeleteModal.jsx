import React from 'react';
import PropTypes from 'prop-types';

function DeleteModal({ handleClose, deleteFunc, feedId, isPost }) {
    return (
        <div>
            {/* The Modal */}
            <div id="myModal" className="modal">
                {/* Modal content */}
                <div className="modal-content">
                    <span onClick={handleClose} className="close">Ã—</span>
                    <p>{`Do you really want to delete this ${isPost ? 'post' : 'comment'}?`}</p>
                    <button onClick={() => deleteFunc(feedId)} className="btn info">Yes</button>
                    <button onClick={handleClose} className="btn default">No</button>
                </div>
            </div>
        </div>
    )
}

DeleteModal.propTypes = {
    handleClose: PropTypes.func,
    deleteFunc: PropTypes.func,
    feedId: PropTypes.string,
    isPost: PropTypes.bool,
}

export default DeleteModal;