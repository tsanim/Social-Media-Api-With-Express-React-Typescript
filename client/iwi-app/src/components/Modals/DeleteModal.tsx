import React from 'react';
import PropTypes from 'prop-types';
import DeleteModalProps from '../../interfaces/Components/Modals/DeleteModalProps.interface';

function DeleteModal({ handleClose, deleteFunc, feedId, isPost }: DeleteModalProps) {
    return (
        <div>
            {/* The Modal */}
            <div id="myModal" className="modal">
                {/* Modal content */}
                <div className="modal-content">
                    <span onClick={handleClose} className="close">Ã—</span>
                    <p>{`Do you really want to delete this ${isPost ? 'post' : 'comment'}?`}</p>
                    <button onClick={() => deleteFunc(feedId)} className="btn info">Yes</button>
                    <button onClick={handleClose} className="btn default">No</button>
                </div>
            </div>
        </div>
    )
}

DeleteModal.propTypes = {
    handleClose: PropTypes.func,
    deleteFunc: PropTypes.func,
    feedId: PropTypes.string,
    isPost: PropTypes.bool,
}

export default DeleteModal;