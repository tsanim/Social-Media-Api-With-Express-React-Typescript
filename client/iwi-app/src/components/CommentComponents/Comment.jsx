import React, { useState, useEffect } from 'react';
import DeleteModal from '../Modals/DeleteModal';
import Modal from '../Modals/Modal';
import UserDataLink from '../UserInfoComponents/UserDataLink';
import PropTypes from 'prop-types';
import CommentContainer from './CommentContainer';
import CommentMeta from './CommentMeta';
import { fetchCommentsLikes } from '../../services/commentsService';

function Comment({ comment, currentUser, likeCommentHandler, dislikeCommentHandler, deleteCommentHandler }) {
    const { creator: commentCreator, text, date, likes, _id } = comment;

    //hook for question modal about deleting
    const [showDeleteModal, setShowModal] = useState(false);
    //hook for showing likes people
    const [showLikesPeopleModal, setShowLikesPeopleModal] = useState(false);
    //hook for handle like post 
    const [likesCount, setlikesCount] = useState(likes.length);
    //hook for handle liked/not liked
    const initCommentIsLiked = likes.includes(localStorage.getItem('userId'))
    const [isLiked, setIsLiked] = useState(initCommentIsLiked);
    //hook for likers
    const [likers, setLikers] = useState([]);

    /*
            TODO: Add AbortController() to useEffect functions to stop request when component is unmount
    */

    //when component did mound , fetched the likers of the comment for showing them in the modal
    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(() => {
        fetchCommentsLikes(_id, (data) => {
            setLikers(() => data.likes);
        });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const handleLikeComment = (e) => {
        setlikesCount((likesCount) => likesCount + 1);
        setIsLiked((isLiked) => !isLiked);
        likeCommentHandler(_id);

        //correct likers like add logged in user in likers array
        setLikers((likers) => [...likers, currentUser]);
    }

    const handleDisLikeComment = (e) => {
        setlikesCount((likesCount) => likesCount - 1);
        setIsLiked((isLiked) => !isLiked);
        dislikeCommentHandler(_id);

        //correct likers like remove logged in user from likers array
        setLikers((likers) => {
            likers = likers.filter(l => l._id !== currentUser._id)
            return likers;
        });
    }

    const handleShowModal = (e) => {
        setShowModal((showDeleteModal) => true);
    }

    const handleShowLikesModal = (e) => {
        setShowLikesPeopleModal((showLikesPeopleModal) => true);
    }

    const handleClose = (e) => {
        setShowModal((showDeleteModal) => false);
        setShowLikesPeopleModal((showLikesPeopleModal) => false);

    }

    return (
        <div className="comment">
            <figure className="commentData">
                <UserDataLink user={commentCreator} />
            </figure>
            <CommentContainer
                text={text}
                isLiked={isLiked}
                handleDisLikeComment={handleDisLikeComment}
                handleLikeComment={handleLikeComment}
                handleShowModal={handleShowModal}
                currentUser={currentUser}
                commentCreator={commentCreator}
            />

            <CommentMeta
                date={date}
                handleShowLikesModal={handleShowLikesModal}
                likesCount={likesCount}
            />

            {
                showDeleteModal
                    ? <DeleteModal
                        isPost={false}
                        feedId={_id}
                        deleteFunc={deleteCommentHandler}
                        handleClose={handleClose} />
                    : null
            }

            {/* Modal for likers */}
            {
                showLikesPeopleModal ? <Modal handleClose={handleClose} modalHeaderName={'Likes'} users={likers} /> : null
            }
        </div >
    )
}

Comment.propTypes = {
    comment: PropTypes.object,
    currentUser: PropTypes.object,
    likeCommentHandler: PropTypes.func,
    dislikeCommentHandler: PropTypes.func,
    deleteCommentHandler: PropTypes.func
}

export default Comment;