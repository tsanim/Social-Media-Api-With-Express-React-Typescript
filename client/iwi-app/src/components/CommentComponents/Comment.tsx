import React, { useState, useEffect } from 'react';
import DeleteModal from '../Modals/DeleteModal';
import Modal from '../Modals/UserModal';
import UserDataLink from '../UserInfoComponents/UserDataLink';
import PropTypes from 'prop-types';
import CommentContainer from './CommentContainer';
import CommentMeta from './CommentMeta';

import CommentsService from '../../services/CommentsService';
import CommentProps from '../../interfaces/Components/CommentComponents/CommentProps.interface';
import { PlainUser } from '../../interfaces/User/User.interface';

function Comment({ comment, currentUser, likeCommentHandler, dislikeCommentHandler, deleteCommentHandler }: CommentProps) {
    const { creator: commentCreator, text, date, likes, _id } = comment;

    //hook for question modal about deleting
    const [showDeleteModal, setShowModal] = useState(false);
    //hook for showing likes people
    const [showLikesPeopleModal, setShowLikesPeopleModal] = useState(false);
    //hook for handle like post 
    const [likesCount, setlikesCount] = useState(likes.length);
    //hook for handle liked/not liked
    const initCommentIsLiked = (likes as string[]).includes(localStorage.getItem('userId'));
    const [isLiked, setIsLiked] = useState(initCommentIsLiked);
    //hook for likers
    const [likers, setLikers] = useState([]);

    /*
            TODO: Add AbortController() to useEffect functions to stop request when component is unmount
    */

    //when component did mound , fetched the likers of the comment for showing them in the modal
    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(() => {
        CommentsService.fetchCommentsLikes(_id, (data) => {
            setLikers(() => data.likes);
        });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const handleLikeComment = () => {
        setlikesCount((likesCount: number) => likesCount + 1);
        setIsLiked((isLiked: boolean) => !isLiked);
        likeCommentHandler(_id);

        //correct likers like add logged in user in likers array
        setLikers((likers: PlainUser[]) => [...likers, currentUser]);
    }

    const handleDisLikeComment = () => {
        setlikesCount((likesCount: number) => likesCount - 1);
        setIsLiked((isLiked: boolean) => !isLiked);
        dislikeCommentHandler(_id);

        //correct likers like remove logged in user from likers array
        setLikers((likers: PlainUser[]) => {
            likers = likers.filter(l => l._id !== currentUser._id)
            return likers;
        });
    }

    const handleShowModal = () => {
        setShowModal(() => true);
    }

    const handleShowLikesModal = () => {
        setShowLikesPeopleModal(() => true);
    }

    const handleClose = () => {
        setShowModal(() => false);
        setShowLikesPeopleModal(() => false);

    }

    return (
        <div className="comment">
            <figure className="commentData">
                <UserDataLink user={commentCreator as PlainUser} />
            </figure>
            <CommentContainer
                text={text}
                isLiked={isLiked}
                handleDislikeComment={handleDisLikeComment}
                handleLikeComment={handleLikeComment}
                handleShowModal={handleShowModal}
                currentUser={currentUser}
                commentCreator={commentCreator as PlainUser}
            />

            <CommentMeta
                date={date}
                handleShowLikesModal={handleShowLikesModal}
                likesCount={likesCount}
            />

            {
                showDeleteModal
                    ? <DeleteModal
                        isPost={false}
                        feedId={_id}
                        deleteFunc={deleteCommentHandler}
                        handleClose={handleClose} />
                    : null
            }

            {/* Modal for likers */}
            {
                showLikesPeopleModal ? <Modal handleClose={handleClose} modalHeaderName={'Likes'} users={likers} /> : null
            }
        </div >
    )
}

Comment.propTypes = {
    comment: PropTypes.object,
    currentUser: PropTypes.object,
    likeCommentHandler: PropTypes.func,
    dislikeCommentHandler: PropTypes.func,
    deleteCommentHandler: PropTypes.func
}

export default Comment;