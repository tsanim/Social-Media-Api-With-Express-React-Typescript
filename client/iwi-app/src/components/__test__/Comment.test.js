import React from 'react';
import Comment from '../CommentComponents/Comment';

import { shallow, mount } from 'enzyme';
import UserDataLink from '../UserInfoComponents/UserDataLink';
import CommentContainer from '../CommentComponents/CommentContainer';
import CommentMeta from '../CommentComponents/CommentMeta';
import DeleteModal from '../Modals/DeleteModal';
import { MemoryRouter } from 'react-router-dom';
import Modal from '../Modals/Modal';

function messageComponentShallowSetup() {
    const props = {
        comment: { likes: ['123'], creator: {}, text: '', date: '', _id: '' },
        currentUser: { _id: '123' },
        likeCommentHandler: jest.fn(),
        dislikeCommentHandler: jest.fn(),
        deleteCommentHandler: jest.fn()
    }

    const enzymeWrapper = shallow(<Comment {...props} />)

    return {
        props,
        enzymeWrapper
    }
}

function messageComponentMountSetup() {
    const props = {
        comment: { likes: ['123'], creator: { _id: '123' }, text: '', date: '', _id: '' },
        currentUser: { _id: '123' },
        likeCommentHandler: jest.fn(),
        dislikeCommentHandler: jest.fn(),
        deleteCommentHandler: jest.fn()
    }

    const enzymeWrapper = mount(
        <MemoryRouter>
            <Comment {...props} />
        </MemoryRouter>
    )

    return {
        props,
        enzymeWrapper
    }
}

describe('Comment component', () => {
    it('should render initial layout', () => {
        const { enzymeWrapper } = messageComponentShallowSetup();

        expect(enzymeWrapper.getElements()).toMatchSnapshot();
    });

    it('should render container div with class "comment"', () => {
        const { enzymeWrapper } = messageComponentShallowSetup();

        expect(enzymeWrapper.find('div.comment')).toHaveLength(1);
    });

    it('should render child figure tag with class "commentData"', () => {
        const { enzymeWrapper } = messageComponentShallowSetup();

        expect(enzymeWrapper.find('figure.commentData')).toHaveLength(1);
    });

    it('should render child UserDataLink Component', () => {
        const { enzymeWrapper } = messageComponentShallowSetup();

        expect(enzymeWrapper.find(UserDataLink)).toHaveLength(1);
    });

    it('should render child CommentContainer Component', () => {
        const { enzymeWrapper } = messageComponentShallowSetup();

        expect(enzymeWrapper.find(CommentContainer)).toHaveLength(1);
    });

    it('should render child CommentMeta component', () => {
        const { enzymeWrapper } = messageComponentShallowSetup();

        expect(enzymeWrapper.find(CommentMeta)).toHaveLength(1);
    });

    it('should render like button tag and update likesCount on button click', () => {
        const { enzymeWrapper, props } = messageComponentMountSetup();

        enzymeWrapper.find('.commentBtns').childAt(0).simulate('click');

        expect(enzymeWrapper.find(CommentMeta).prop('likesCount')).toEqual(props.comment.likes.length + 1);
    });


    it('should render button tag with class "liked" and update likesCount on click', () => {
        window.localStorage.setItem('userId', '123');

        const { enzymeWrapper, props } = messageComponentMountSetup();

        enzymeWrapper.find('.commentBtns button.liked').simulate('click');

        expect(enzymeWrapper.find(CommentMeta).prop('likesCount')).toEqual(props.comment.likes.length - 1);
    });

    it('should render Modal Component when span tag for likers is clicked', () => {
        const { enzymeWrapper } = messageComponentMountSetup();

        //find second span child of meta div and simulate click
        enzymeWrapper.find('div.meta').childAt(1).simulate('click');

        expect(enzymeWrapper.find(Modal)).toHaveLength(1);
    });

    it('should render two buttons tag when comment creator id is equal to currentUser id', () => {
        const { enzymeWrapper } = messageComponentMountSetup();

        expect(enzymeWrapper.find('.commentBtns').children()).toHaveLength(2);
    });

    it('should render DeleteModal Component when delete button is clicked', () => {
        const { enzymeWrapper } = messageComponentMountSetup();

        enzymeWrapper.find('.commentBtns').childAt(1).simulate('click');

        expect(enzymeWrapper.find(DeleteModal)).toHaveLength(1);
    });

    it('should unmount DeleteModal Component when close button is clicked', () => {
        const { enzymeWrapper } = messageComponentMountSetup();

        enzymeWrapper.find('.commentBtns').childAt(1).simulate('click');

        expect(enzymeWrapper.find(DeleteModal)).toHaveLength(1);

        enzymeWrapper.find('.modal-content span.close').simulate('click');

        expect(enzymeWrapper.find(DeleteModal)).toHaveLength(0);
    });
});