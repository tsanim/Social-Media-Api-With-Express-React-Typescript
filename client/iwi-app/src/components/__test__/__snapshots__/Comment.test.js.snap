// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment component should render initial layout 1`] = `
Array [
  <div
    className="comment"
  >
    <figure
      className="commentData"
    >
      <UserDataLink
        user={Object {}}
      />
    </figure>
    <CommentContainer
      commentCreator={Object {}}
      currentUser={
        Object {
          "_id": "123",
        }
      }
      handleDisLikeComment={[Function]}
      handleLikeComment={[Function]}
      handleShowModal={[Function]}
      isLiked={false}
      text=""
    />
    <CommentMeta
      date=""
      handleShowLikesModal={[Function]}
      likesCount={1}
    />
  </div>,
]
`;
